qutip/core/data/tidyup.pyx

```diff
@@ -39,3 +44,53 @@ cpdef CSR tidyup_csr(CSR matrix, double tol, bint inplace=True):
     if nnz_new == nnz_orig:
         return out
     return out
+
+
+from .dispatch import Dispatcher as _Dispatcher
+import inspect as _inspect
+
+# In this case, to support the `inplace` argument, we do not support
+# dispatching on the output.
+
+tidyup = _Dispatcher(
+    _inspect.Signature([
+        _inspect.Parameter('matrix', _inspect.Parameter.POSITIONAL_OR_KEYWORD),
+        _inspect.Parameter('tol', _inspect.Parameter.POSITIONAL_OR_KEYWORD),
+        _inspect.Parameter('inplace', _inspect.Parameter.POSITIONAL_OR_KEYWORD,
+                           default=True),
+    ]),
+    name='tidyup',
+    module=__name__,
+    inputs=('matrix',),
+    out=False,
+)
+tidyup.__doc__ =\
+    """
+    Tidy up the input matrix by truncating small values to zero.  The real and
+    imaginary parts are treated individually, so (for example) the number
+        1e-18 + 2j
+    would be truncated with a tolerance of `1e-15` to just
+        2j
+
+    By default, this operation is in-place.  The output type will always match
+    the input type; no dispatching takes place on the output.
+
+    Arguments
+    ---------
+    matrix : Data
+        The matrix to tidy up.
+
+    tol : real
+        The absolute tolerance to use to determine whether a real or imaginary
+        part should be truncated to zero.
+
+    inplace : bool, optional (True)
+        Whether to do the operation in-place.  The output matrix will always be
+        returned, even if this argument is `True`; it will just be the same
+        Python object as was input.
+    """
+tidyup.add_specialisations([
+    (CSR, tidyup_csr),
```

Comment:
Well spotted!  The new PR I'll send soon makes some steps to addressing this, but I didn't want to overburden this one.

We do need to add a lot more documentation and make the `auto_tidyup` option _much_ more visible, because on small sizes, it actually adds a lot of overhead to mathematical operations.  With regards to the dispatchers, this problem is perhaps solved a little better by having the default method of resolving unary operations (like `tidyup`) when no output type is specified being to maintain the same type, even if that means converting the output back.

I imagine there's loads of little tweaks and usage papercuts to find and solve that a beta period will really help uncover.

Reply:
By "makes some steps to addressing this", I actually just mean that I added a `tidyup_dense` method.

This PR actually misrepresents all that's required to make a new type at the moment; currently you also have to specify `adjoint`, `conj` and `transpose` methods, but I'll likely remove that requirement.  At any rate, once I've written the proper documentation for how to add a new type, there'll be a full guide on how to do it, with everything that needs to be done.

I think I'm going to move the relevant methods onto the `data.Data` ABC, and now I've solved the circular dispatch import problem, I'll likely attach the full dispatchers in as the mathematical methods, and the `adjoint`, `transpose`, etc methods.

Reply:
See also: #1349.